ring/batch.go

```diff
@@ -111,30 +128,56 @@ func DoBatch(ctx context.Context, op Operation, r ReadRing, keys []uint32, callb
 	}
 }
 
-func (b *batchTracker) record(sampleTrackers []*itemTracker, err error) {
-	// If we succeed, decrement each sample's pending count by one.  If we reach
-	// the required number of successful puts on this sample, then decrement the
-	// number of pending samples by one.  If we successfully push all samples to
-	// min success instances, wake up the waiting rpc so it can return early.
-	// Similarly, track the number of errors, and if it exceeds maxFailures
-	// shortcut the waiting rpc.
+func (b *batchTracker) record(itemTrackers []*itemTracker, err error) {
+	// If we reach the required number of successful puts on this item, then decrement the
+	// number of pending items by one.
 	//
-	// The use of atomic increments here guarantees only a single sendSamples
-	// goroutine will write to either channel.
-	for i := range sampleTrackers {
+	// The use of atomic increments here is needed as:
+	// * rpcsPending and rpcsFailed guarantees only a single goroutine will write to either channel
+	// * succeeded, failed4xx, failed5xx and remaining guarantees that the "return decision" is made atomically
```

Comment:
[Nit] Language improvements:
```suggestion
	// * rpcsPending and rpcsFailed guarantee only a single goroutine will write to either channel
	// * succeeded, failed4xx, failed5xx and remaining guarantee that the "return decision" is made atomically
```